[Metrics v2] Add the ability to edit a Metric definition page

nemanjaglumac opened the pull request:
This PR introduces the "edit query" page for metrics.
The new route it uses is `/metric/:slug/query`.

This page is in many ways analogous to the existing "edit query" page for models.
One notable exception that we needed to make in this PR is to [prevent the page from fetching model indexes](https://github.com/metabase/metabase/pull/38664/commits/f9d2c9af717d9e768817f27c507364cbd5efc314).

As we add more pages, and as metrics find their way into different parts of our app, it's becoming evident that the old abstractions (and the old naming patterns) that were tailored specifically to models are insufficient. A lot of what we're doing to make metrics work is patching that old code. And it most cases, the crux of the change is going from `isModel` to `isModelOrMetric`.

### How to verify these changes work?
1. Open the existing metric
2. Click on the `...` menu
3. You should see "Edit query definition", and you should be able to click on it
4. After clicking on it, you should lang on `/metric/:slug/query` page

Please note that, unlike for models, we're currently NOT displaying "query" and "metadata" tabs. There are separate tasks for that in step 7 of the parent epic.

Closes #37354

nemanjaglumac committed:
Add new route for `/metric/:id/query`

nemanjaglumac committed:
Expose "edit query" option in the menu

The "edit metadata" option stays hidden for now.

nemanjaglumac committed:
Prevent fetching model indexes for metrics

nemanjaglumac committed:
Account for the metrics when computing the dirty state

kamilmielnik commented on frontend/src/metabase/query_builder/components/view/ViewHeader/components/QuestionActions/QuestionActions.tsx:
```diff
@@ -91,18 +91,19 @@ export const QuestionActions = ({
   const isSaved = question.isSaved();
   const database = question.database();
   const canAppend = canUpload && canWrite && !!question._card.based_on_upload;
   const type = question.type() || "question";
```
> @nemanjaglumac could you modify `question.type()` to default to `"question"` and remove default values from places where it is called?

I did this in #38714, feel free to skip it.

nemanjaglumac replied:
I can, for sure.
@kamilmielnik just double checking that this does not clash with the work you're doing?

nemanjaglumac replied:
Awesome. Thanks

nemanjaglumac committed:
Fix card type

nemanjaglumac committed:
Account for potentially undefined `originalQuestion`

nemanjaglumac committed:
Fix and expand the `DatasetEditor` unit tests

nemanjaglumac committed:
Fix `QuestionActions` unit tests

nemanjaglumac committed:
Fix the logic in the `getIsDirty` selector call stack

nemanjaglumac committed:
Introduce `useModelIndexesListQuery` hook

nemanjaglumac committed:
Use new `useModelIndexesListQuery` hook in `DatasetEditor`

nemanjaglumac commented on frontend/src/metabase-types/api/modelIndexes.ts:
```diff
@@ -25,3 +25,9 @@ export type IndexedEntity = {
+  model_id: CardId | null;
+};
+
+export type ModelIndexesListResult = ModelIndex[];
```
I think the shape of the response as defined in `ModelIndex` became outdated over time. If you observe the response in dev tools or using any API tool, you'll see that `generation` and `state_changed_at` don't exist.

Two new keys I see in the response are:
- `created_at` and
- `indexed_at`
(both are datetime strings)

Here's the example response:
```json
[
	{
		"value_ref": [
			"field",
			71,
			null
		],
		"schedule": "0 0 4 * * ? *",
		"pk_ref": [
			"field",
			68,
			null
		],
		"state": "indexed",
		"creator_id": 1,
		"model_id": 78,
		"id": 3,
		"indexed_at": "2024-02-15T10:12:08.806019+01:00",
		"error": null,
		"created_at": "2024-02-15T10:12:08.664025+01:00"
	}
]
```

@ranquild @kamil (and @iethree ) should this type be updated?
If this new hook proves to be working (and that's a huge if), should we maybe introduce it to `master` and update types? Also, should we remove the current conditional logic introduced [here](https://github.com/metabase/metabase/pull/37284/files#diff-24b8c743de25248f2f33f660f56dd21dce18044275e746374b1b6696e27d8868L22-R24) and use the hook instead?

nemanjaglumac replied:
It seems cleaner to (1) open a PR with the new model-indexes hook against master and to merge it there. After that, (2) merge master back into `metrics-v2` and (3) rebase this branch on top of the updated metrics-v2.

That way this PR would not even touch model-indexes code, which would make the diff much more straightforward.

Either way is fine with me. Just listing options here, but will do whatever reviewers think is best.

iethree replied:
FWIW, model-indexes shouldn't be an entity - it's just that our data fetching/caching options are kind of limited, it was easiest to try to shove it into that framework. We're currently working establishing some nice easy patterns for fetching and caching non-entities that should make this a lot cleaner.

In the mean time, yes, absolutely update the API types in master if they no longer reflect the reality of what the backend is serving. I'd double check with a BE engineer to see what happened to `generation` and `state_changed_at`

nemanjaglumac committed:
Fix type check failure

This commit remove the superfluous `query` prop from the `DatasetEditor`
component in the related unit test.

nemanjaglumac committed:
Handle `modelIndexes` prop potentially being undefined

This commit fixes failing E2E tests due to the broken `fieldHasIndex` util.
